// pdf2xml/lib/pdftotext.js
var shell = require('shelljs');

// we can only do like this because of a bug in shelljs
// global silent
shell.config.silent = true;

var xmlOptions = ['-xml', '-i', '-stdout'];

function pdftotext (filename, options) {
  this.options = options || {};
  //quote filename
  this.options.additional = ['"' + filename + '"'];

  pdftotext.prototype.add_options = function(optionArray) {
    if (typeof optionArray.length !== undefined) {
        var self = this;
        optionArray.forEach(function(el) {
          if (el.indexOf(' ') > 0) {
            var values = el.split(' ');
            self.options.additional.push(values[0], values[1]);
          } else {
            self.options.additional.push(el);
          }
        });
    }
    return this;
  };

  pdftotext.prototype.getXMLSync = function() {
    var self = this;
    self.add_options(xmlOptions);
    var child = shell.exec('pdftohtml ' + self.options.additional.join(' '));
    if (child.code === 0) {
      return child.output;
    }
    else {
      if (!shell.which('pdftohtml')) {
        throw new Error('pdftohtml (poppler-utils) is missing. Hint: sudo apt-get install poppler-utils');
      }
      throw new Error("pdftohtml error: "+ child.output);
    }
  }

  pdftotext.prototype.getXML = function(cb) {
    var self = this;
    self.add_options(xmlOptions);
    var child = shell.exec('pdftohtml ' + self.options.additional.join(' '), function(code, data) {
      if (code === 0) {
        if (cb && typeof cb === "function") {
          cb(null, data, self.options.additional);
        }
      }
      else {
        var err;
        if (!shell.which('pdftohtml')) {
          err = new Error('pdftohtml (poppler-utils) is missing. Hint: sudo apt-get install poppler-utils');
        }
        else {
          err = new Error(data);
        }
        if (cb && typeof cb === "function") {
          cb(err, data, self.options.addtional);
        }
      }
    });
  }

  pdftotext.prototype.error = function(callback) {
    this.options.error = callback;
    return this;
  };

  pdftotext.prototype.success = function(callback) {
    this.options.success = callback;
    return this;
  };
}

// module exports
exports = module.exports = function(filename, args) {
  return new pdftotext(filename, args);
};
